File: src/frontend/nfe/lib/app/data/model/nfe_nf_referenciada_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeNfReferenciadaModel {
	int? id;
	int? idNfeCabecalho;
	int? codigoUf;
	String? anoMes;
	String? cnpj;
	String? modelo;
	String? serie;
	int? numeroNf;

	NfeNfReferenciadaModel({
		this.id,
		this.idNfeCabecalho,
		this.codigoUf,
		this.anoMes,
		this.cnpj,
		this.modelo,
		this.serie,
		this.numeroNf,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo_uf',
		'ano_mes',
		'cnpj',
		'modelo',
		'serie',
		'numero_nf',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo Uf',
		'Ano Mes',
		'Cnpj',
		'Modelo',
		'Serie',
		'Numero Nf',
	];

	NfeNfReferenciadaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCabecalho = jsonData['idNfeCabecalho'];
		codigoUf = jsonData['codigoUf'];
		anoMes = jsonData['anoMes'];
		cnpj = jsonData['cnpj'];
		modelo = NfeNfReferenciadaDomain.getModelo(jsonData['modelo']);
		serie = NfeNfReferenciadaDomain.getSerie(jsonData['serie']);
		numeroNf = jsonData['numeroNf'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCabecalho'] = idNfeCabecalho != 0 ? idNfeCabecalho : null;
		jsonData['codigoUf'] = codigoUf;
		jsonData['anoMes'] = anoMes;
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['modelo'] = NfeNfReferenciadaDomain.setModelo(modelo);
		jsonData['serie'] = NfeNfReferenciadaDomain.setSerie(serie);
		jsonData['numeroNf'] = numeroNf;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCabecalho = plutoRow.cells['idNfeCabecalho']?.value;
		codigoUf = plutoRow.cells['codigoUf']?.value;
		anoMes = plutoRow.cells['anoMes']?.value;
		cnpj = plutoRow.cells['cnpj']?.value;
		modelo = plutoRow.cells['modelo']?.value != '' ? plutoRow.cells['modelo']?.value : 'AAA';
		serie = plutoRow.cells['serie']?.value != '' ? plutoRow.cells['serie']?.value : 'AAA';
		numeroNf = plutoRow.cells['numeroNf']?.value;
	}	

	NfeNfReferenciadaModel clone() {
		return NfeNfReferenciadaModel(
			id: id,
			idNfeCabecalho: idNfeCabecalho,
			codigoUf: codigoUf,
			anoMes: anoMes,
			cnpj: cnpj,
			modelo: modelo,
			serie: serie,
			numeroNf: numeroNf,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_nf_referenciada_model.py
from src import db


class NfeNfReferenciadaModel(db.Model):
    __tablename__ = 'nfe_nf_referenciada'

    id = db.Column(db.Integer, primary_key=True)
    codigo_uf = db.Column(db.Integer)
    ano_mes = db.Column(db.String(4))
    cnpj = db.Column(db.String(14))
    modelo = db.Column(db.String(2))
    serie = db.Column(db.String(3))
    numero_nf = db.Column(db.Integer)
    id_nfe_cabecalho = db.Column(db.Integer, db.ForeignKey('nfe_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cabecalho = data.get('idNfeCabecalho')
        self.codigo_uf = data.get('codigoUf')
        self.ano_mes = data.get('anoMes')
        self.cnpj = data.get('cnpj')
        self.modelo = data.get('modelo')
        self.serie = data.get('serie')
        self.numero_nf = data.get('numeroNf')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCabecalho': self.id_nfe_cabecalho,
            'codigoUf': self.codigo_uf,
            'anoMes': self.ano_mes,
            'cnpj': self.cnpj,
            'modelo': self.modelo,
            'serie': self.serie,
            'numeroNf': self.numero_nf,
        }